public/app/features/explore/QueryRows.tsx

```diff
@@ -21,10 +21,11 @@ export default class QueryRows extends PureComponent<QueryRowsProps> {
     const { className = '', exploreEvents, exploreId, initialQueries } = this.props;
     return (
       <div className={className}>
-        {initialQueries.map((query, index) => (
-          // TODO instead of relying on initialQueries, move to react key list in redux
-          <QueryRow key={query.key} exploreEvents={exploreEvents} exploreId={exploreId} index={index} />
-        ))}
+        {initialQueries.map((query, index) => {
+          // using query.key will introduce infinite loop because QueryEditor#53
+          const key = query.datasource ? `${query.datasource}-${index}` : query.key;
```

Comment:
When switching from Prom or Loki datasource to gdev for instance, we end up in an infinite loop because QueryEditor calls (line 53), so if we use query.key here the key will be regenerated every time the QueryEditor is unmounted and mounted again.
```typescript 
this.props.onQueryChange(target);
```